Replace unsound struct-cast simplification

smowton opened the pull request:
Previously the simplifier would simplify ((struct A)x) into (x) if x.type() and A
were structurally compatible (i.e. had the same named and typed fields). However this
isn't really a sound simplification, as it changes the type of the expression. I therefore
replace the rule with one that only applies within the common-case member expression, i.e.
simplifiying ((struct A)x).y -> x.y, which does not change the type of the outer expression.

Notes: @allredj @romainbrenguier initially I simply disabled the unsound simplification, however this caused two string-smoke-tests to fail because a hard `java.lang.String` -> `java.lang.CharSequence` cast went unsimplified, preventing constant propagation from completing and causing the `contains` and `append_string` tests to fail because they needed to use their general case rather than the special case where a parameter is known to be a constant. If you try disabling the new simplification introduced here you should be able to reproduce the problem -- this might be an interesting bug, or could just be a known existing weakness in the solver.

This solution could be generalised by permitting this simplification *only for inner expressions*, i.e. casts which do not occur at the root of the expression tree. @tautschnig do we want to try that, or is handling this special case enough?

smowton committed:
Replace unsound struct-cast simplification

Previously the simplifier would simplify ((struct A)x) into (x) if x.type() and A
were structurally compatible (i.e. had the same named and typed fields). However this
isn't really a sound simplification, as it changes the type of the expression. I therefore
replace the rule with one that only applies within the common-case member expression, i.e.
simplifiying ((struct A)x).y -> x.y, which does not change the type of the outer expression.

tautschnig commented:
We had that discussion elsewhere - would you happen to have pointers to that?

tautschnig commented on src/util/simplify_expr_struct.cpp:
```diff
@@ -208,6 +209,17 @@ bool simplify_exprt::simplify_member(exprt &expr)
+    if(base_type_eq(op_type, op.op0().type(), ns))
+    {
+      expr.op0() = op.op0();
+      simplify_rec(expr);
```
I don't think it's necessary to simplify recursively, I'd say a call to `simplify_member` should do the trick.

tautschnig replied:
I think this is the only bit keeping me from approving.

tautschnig commented:
It would be great to have some regression tests that show that the previous one was unsound. Otherwise I'm fine with this special case.

smowton commented:
@tautschnig for reference the previous discussion of the same problem is here: https://github.com/diffblue/cbmc/pull/1499#issuecomment-346793910

smowton committed:
String smoke tests: ensure no type mismatches are seen

Type mismatches previously resulted from the expression simplifier removing casts
from one struct type to another, so e.g. (struct A)b would be simplified to b,
changing its type and leading to an equality with a type mismatch. Now that the
simplifier has been fixed this should no longer happen -- this change ensures that
none of the smoke tests generate the warning any longer.

smowton commented:
@tautschnig the string smoke tests were exposing the problem, so I've added an extra commit to verify they don't any longer. Tagging string developers to check this is reasonable.

I've also implemented your comment above.

tautschnig commented on regression/strings-smoke-tests/java_append_object/test.desc:
```diff
@@ -1,8 +1,9 @@
 ^VERIFICATION SUCCESSFUL$
 --
+non equal types
 Issue: diffblue/test-gen#82
```
There is a problem in the original test description file: this comment "Issue: ..." should have a preceding `--` line to separate strings-not-to-be-found-in-output from comments.

tautschnig commented on regression/strings-smoke-tests/java_char_array_init/test.desc:
```diff
@@ -1,8 +1,9 @@
 ^VERIFICATION SUCCESSFUL$
 --
+non equal types
 cbmc/test-gen#259
```
As above, the comment needs a preceding `--` line.

tautschnig commented on regression/strings-smoke-tests/java_equal/test_2.desc:
```diff
@@ -1,7 +1,8 @@
 ^SIGNAL=0$
 --
+non equal types
 https://github.com/diffblue/test-gen/issues/856
```
As above: `--` required.

tautschnig commented on regression/strings-smoke-tests/java_if/test.desc:
```diff
@@ -1,9 +1,10 @@
 ^\[.*assertion.2\].* line 13.* FAILURE$
 --
+non equal types
 $ignoring\s*char\s*array
```
Why does this pattern start with a `$`?

tautschnig commented on regression/strings-smoke-tests/java_index_of_char/test.desc:
```diff
@@ -1,8 +1,9 @@
 ^VERIFICATION SUCCESSFUL$
 --
+non equal types
 Issue: cbmc/test-gen#77
```
As above: `--` required.

tautschnig commented on regression/strings-smoke-tests/java_int_to_string/test2_bug.desc:
```diff
@@ -1,11 +1,12 @@
+non equal types
 --
 When string-max-length is not set, the solver can run out of memory in
 an unpredictable way.
```
Missing newline at end-of-file.

tautschnig commented on regression/strings-smoke-tests/java_length2/test.desc:
```diff
@@ -1,10 +1,11 @@
 ^VERIFICATION SUCCESSFUL$
 --
+non equal types
 This test currently fails because we do not handle the case where the argument
```
As above: `--` needed

tautschnig commented on regression/strings-smoke-tests/java_replace/test.desc:
```diff
@@ -1,10 +1,11 @@
 ^VERIFICATION SUCCESSFUL$
 --
+non equal types
 diffblue/test-gen#256
```
As above: `--` required

tautschnig reviewed (approved):
Looks good, but the *.desc files need some (mostly trivial) work.

smowton committed:
Fix unsatisfiable test line

smowton committed:
Fix missing newline at end of desc file

smowton committed:
Fix testing for empty line in test desc file

smowton commented:
@tautschnig changes addressed. Looks like my script to add the rejected lines wasn't quite smart enough :)

romainbrenguier reviewed (approved):
this looks fine
